syntax = "proto3";

package orderbook;

import "OrderEntry.proto";  // Import Side enum from OrderEntry

// Market data streaming service
service MarketData {
    rpc Subscribe (SubscribeRequest) returns (stream MarketUpdate);
}

message SubscribeRequest {
    string symbol = 1;
}

// Updated to use snake_case for generated code
message MarketUpdate {
    uint64 seq = 1;
    string symbol = 2;
    
    oneof update_type {
        BookUpdate book_update = 3;
        TradeUpdate trade_update = 4;
    }
}

message BookUpdate {
    double best_bid = 1;
    double best_ask = 2;
    uint32 bid_size = 3;
    uint32 ask_size = 4;
}

message TradeUpdate {
    double price = 1;
    uint32 quantity = 2;
    Side taker_side = 3;
}
